File: backend/scripts/migrate.py
"""Database migration script for FastAPI backend.

This script runs Alembic migrations programmatically with enhanced error handling
and logging. It provides better visibility into the migration process and ensures
consistent behavior across different deployment environments.
"""

import sys
from pathlib import Path

from alembic import command
from alembic.config import Config


def run_migrations() -> None:
    """Run database migrations using Alembic programmatic API."""
    print("🔄 Starting backend database migrations...")

    try:
        # Get the backend directory (parent of scripts directory)
        backend_dir = Path(__file__).parent.parent

        # Create Alembic config
        alembic_ini_path = backend_dir / "alembic.ini"

        if not alembic_ini_path.exists():
            print(f"❌ Alembic configuration file not found: {alembic_ini_path}")
            sys.exit(1)

        # Load Alembic configuration
        alembic_cfg = Config(str(alembic_ini_path))

        # Set the script location to the alembic directory
        alembic_cfg.set_main_option("script_location", str(backend_dir / "alembic"))

        # Run migrations to head
        command.upgrade(alembic_cfg, "head")

        print("✅ Backend database migrations completed successfully")
        sys.exit(0)

    except Exception as error:
        print(f"❌ Migration failed: {error}")
        sys.exit(1)


if __name__ == "__main__":
    run_migrations()

File: nextjs/scripts/migrate.ts
import { resolve } from "node:path";
import { config } from "dotenv";
import { drizzle } from "drizzle-orm/postgres-js";
import { migrate } from "drizzle-orm/postgres-js/migrator";
import postgres from "postgres";

// Load .env from project root directory
config({ path: resolve(__dirname, "../../.env") });

const DATABASE_URL = process.env.DATABASE_URL;

if (!DATABASE_URL) {
  console.error("❌ DATABASE_URL environment variable is not set");
  process.exit(1);
}

// Convert postgresql+asyncpg:// to postgresql:// for postgres-js
const databaseUrl = DATABASE_URL.replace("+asyncpg", "");

async function runMigrations() {
  console.log("🔄 Starting database migrations...");

  // Create postgres connection for migrations
  const migrationClient = postgres(databaseUrl, { max: 1 });

  try {
    // Run migrations
    await migrate(drizzle(migrationClient), {
      migrationsFolder: resolve(__dirname, "../drizzle"),
    });

    console.log("✅ Database migrations completed successfully");
    process.exit(0);
  } catch (error) {
    console.error("❌ Migration failed:", error);
    process.exit(1);
  } finally {
    // Close the connection
    await migrationClient.end();
  }
}

// Run migrations
runMigrations();
